fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneRepartitionPlanner.java

```diff
@@ -0,0 +1,170 @@
+        repartitioningContext.sourcePartition = candidatePartition;
+
+        if (currentPartitionCount < indexPartitionLowWatermark) {
+            if (currentPartitionCount > repartitionDocumentCount) {
+                repartitioningContext.countToMove = repartitionDocumentCount + 1;
+                repartitioningContext.newBoundaryRecordPresent = true;
+            } else {
+                repartitioningContext.countToMove = currentPartitionCount;
+                repartitioningContext.emptyingPartition = true;
+            }
+
+            if (olderPartition != null && olderPartition.getCount() + currentPartitionCount <= indexPartitionHighWatermark) {
+                repartitioningContext.action = RepartitioningAction.MERGE_INTO_OLDER;
+            } else if (newerPartition != null && newerPartition.getCount() + currentPartitionCount <= indexPartitionHighWatermark) {
+                repartitioningContext.action = RepartitioningAction.MERGE_INTO_NEWER;
+            } else if (olderPartition != null && newerPartition != null && (2 * indexPartitionHighWatermark - olderPartition.getCount() - newerPartition.getCount()) >= currentPartitionCount ) {
+                repartitioningContext.action = RepartitioningAction.MERGE_INTO_BOTH;
+                // split current and merge into newerPartition and olderPartition.
+                // for this iteration, we do a partial move into olderPartition; next iteration(s) would eventually
+                // move the rest to newerPartition
+                int olderPartitionCapacity = indexPartitionHighWatermark - olderPartition.getCount();
+                repartitioningContext.countToMove = Math.min(olderPartitionCapacity, repartitionDocumentCount);
+                if (currentPartitionCount > repartitioningContext.countToMove) {
+                    repartitioningContext.countToMove++;
+                    repartitioningContext.newBoundaryRecordPresent = true;
+                }
+                repartitioningContext.emptyingPartition = false;
+            } else {
+                repartitioningContext.action = RepartitioningAction.NO_CAPACITY_FOR_MERGE;
+            }
+        } else {
+            // partition is above high watermark
+            repartitioningContext.countToMove = 1 + Math.min(repartitionDocumentCount, indexPartitionHighWatermark);
+            // sanity: in case someone passes a high enough repartitionDocumentCount such that removing
+            // this many docs would flip the partition from being above the high watermark to being below
+            // the low watermark
+            if (currentPartitionCount - repartitioningContext.countToMove < indexPartitionLowWatermark) {
+                repartitioningContext.countToMove = (indexPartitionHighWatermark - indexPartitionLowWatermark) / 2;
+            }
+            repartitioningContext.newBoundaryRecordPresent = true;
+            repartitioningContext.action = RepartitioningAction.OVERFLOW;
```

Comment:
I don't think we should change it now, but it might make sense to put a bit more into this class and have it differentiate between inserting into the `olderPartition` and creating a new partition in between.